packages/xl-ai/src/prosemirror/agent.ts

```diff
@@ -0,0 +1,325 @@
+import { UnreachableCaseError } from "@blocknote/core";
+import { Fragment, Node, Schema, Slice } from "prosemirror-model";
+import { AllSelection, TextSelection, Transaction } from "prosemirror-state";
+import {
+  ReplaceAroundStep,
+  ReplaceStep,
+  Step,
+  Transform,
+} from "prosemirror-transform";
+
+export type AgentStep = {
+  prosemirrorSteps: Step[];
+  selection:
+    | {
+        anchor: number;
+        head: number;
+      }
+    | undefined;
+  type: "select" | "replace" | "insert";
+};
+
+/**
+ * Takes an array of ReplaceSteps and applies them as a human-like agent.
+ * Every step is split into 3 phases:
+ * - select the text to be replaced (1 transaction per ReplaceStep)
+ * - replace the text with the first character of the replacement (if any) (1 transaction per ReplaceStep)
+ * - insert the replacement character by character (strlen-1 transactions per ReplaceStep)
+ */
+export function getStepsAsAgent(doc: Node, pmSchema: Schema, steps: Step[]) {
+  const { modification } = pmSchema.marks;
+
+  const agentSteps: AgentStep[] = [];
+
+  const tr = new Transform(doc);
+
+  for (const step of steps) {
+    if ((step as any).structure) {
+      // Note: for structure changes (e.g.: node type or attr changes) we currently
+      // don't trigger any selection
+
+      if (step instanceof ReplaceStep) {
+        if (
+          step.to !== step.from + 1 ||
+          step.slice.openStart !== 0 ||
+          step.slice.openEnd !== 1 ||
+          step.slice.content.size !== 2
+        ) {
+          throw new Error(
+            "Structure change is not in expected format (ReplaceStep)",
+          );
+        }
+      } else if (step instanceof ReplaceAroundStep) {
+        // we only support ReplaceAroundSteps generated by SetNodeMarkup
+        if (
+          step.insert !== 1 ||
+          step.slice.size !== 2 ||
+          step.gapTo !== step.to - 1 ||
+          step.gapFrom !== step.from + 1
+        ) {
+          throw new Error(
+            "Structure change is not in expected format (ReplaceAroundStep)",
+          );
+        }
+      } else {
+        throw new Error("Step is not a ReplaceStep or ReplaceAroundStep");
+      }
+      const newNode = step.slice.content.firstChild!;
+      const oldNode = tr.doc.resolve(tr.mapping.map(step.from)).nodeAfter!;
+
+      let marks = newNode.marks || [];
+
+      if (newNode.type !== oldNode.type) {
+        marks = modification
+          .create({
+            type: "nodeType",
+            previousValue: oldNode.type.name,
+            newValue: newNode.type.name,
+          })
+          .addToSet(marks);
+      }
+
+      const attrNames = new Set([
+        ...Object.keys(newNode.attrs),
+        ...Object.keys(oldNode.attrs),
+      ]);
+      for (const attr of attrNames) {
+        if (newNode.attrs[attr] !== oldNode.attrs[attr]) {
```

Comment:
attributes can be objects in prosemirror, though this works for blocknote attributes

Reply:
Good to know. I think `prosemirror-suggest-changes` also only supports string attrs fwiw